cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,329 @@
+ * equals `LEFT_TABLE`; otherwise, from `source_right_mask`.
+ *
+ * `source_left_mask`, `source_right_mask` and `destination_mask` must not
+ * overlap.
+ *
+ * @param[in] source_left_mask The left mask whose bits will be merged
+ * @param[in] source_right_mask The right mask whose bits will be merged
+ * @param[out] destination_mask The output mask after merging the left and right masks
+ * @param[in] num_destination_rows The number of bits in the destination_mask
+ * @param[in] merged_table_indices The map that indicates from which input mask bits
+ * will be copied to the output. Length must be equal to `num_destination_rows`
+ * @param[in] merged_row_indices The map that indicates which bit from the input
+ * mask (indicated by `merged_table_indices`) will be copied to the output. Length
+ * must be equal to `num_destination_rows`
+ */
+__global__ void materialize_merged_bitmask_kernel(
+    bit_mask_t const* const __restrict__ source_left_mask,
+    bit_mask_t const* const __restrict__ source_right_mask,
+    bit_mask_t* const destination_mask,
+    gdf_size_type const num_destination_rows,
+    gdf_size_type const* const __restrict__ merged_table_indices,
+    gdf_size_type const* const __restrict__ merged_row_indices) {
+
+  gdf_index_type destination_row = threadIdx.x + blockIdx.x * blockDim.x;
+
+  auto active_threads =
+      __ballot_sync(0xffffffff, destination_row < num_destination_rows);
+
+  while (destination_row < num_destination_rows) {
+    bit_mask_t const* source_mask = merged_table_indices[destination_row] == LEFT_TABLE ? source_left_mask : source_right_mask;
+    bool const source_bit_is_valid{
+        source_mask ? bit_mask::is_valid(source_mask, merged_row_indices[destination_row]) : true};
+    
+    // Use ballot to find all valid bits in this warp and create the output
+    // bitmask element
+    bit_mask_t const result_mask{
+        __ballot_sync(active_threads, source_bit_is_valid)};
+
+    gdf_index_type const output_element = cudf::util::detail::bit_container_index<bit_mask_t, gdf_index_type>(destination_row);
+    
+    // Only one thread writes output
+    if (0 == threadIdx.x % warpSize) {
+      destination_mask[output_element] = result_mask;
+    }
+
+    destination_row += blockDim.x * gridDim.x;
+    active_threads =
+        __ballot_sync(active_threads, destination_row < num_destination_rows);
+  }
+}
+
+void materialize_bitmask(gdf_column const* left_col,
+                        gdf_column const* right_col,
+                        gdf_column* out_col,
+                        gdf_size_type const* table_indices,
+                        gdf_size_type const* row_indices,
+                        cudaStream_t stream) {
+    constexpr gdf_size_type BLOCK_SIZE{256};
+    cudf::util::cuda::grid_config_1d grid_config {out_col->size, BLOCK_SIZE };
+    // TODO: cant use just one template flag for valids, so need to test for null explicitly
```

Comment:
Why not? Since you're invoking a kernel per column, you should be able to just test each column for the presence of null values and then invoke the proper instantiation.

Reply:
The kernel receives two bit masks so there would be actually four cases. No sure if it's OK to specialize for all of them. Or make the kernel just accept one bit mask and launch it two times?

Reply:
I'd just use two template bool parameters. I very much doubt the compile time will suffer significantly. For example, see what I did here for groupby: https://github.com/rapidsai/cudf/blob/branch-0.9/cpp/src/groupby/hash/groupby.cu#L408